dali/sasha/saarch.cpp

```diff
@@ -376,6 +376,269 @@ void WUiterate(ISashaCommand *cmd, const char *mask)
+            CIArrayOf<CDirectoryEntry> sortedDirs;
+            Owned<IDirectoryIterator> dirIterator = dir->directoryFiles(dirMask.isEmpty() ? nullptr : dirMask.str(), false, true);
+            sortDirectory(sortedDirs, *dirIterator, SD_bynameNC, sortInc, true); 
+            ForEachItemIn(sortedDirsIdx, sortedDirs)
+            {
+                CDirectoryEntry &dir = sortedDirs.item(sortedDirsIdx);
+                if (!dir.isdir || !checkDirs(dir))
+                    continue;
+
+                CIArrayOf<CDirectoryEntry> sortedFiles;
+                Owned<IDirectoryIterator> fileIterator = dir.file->directoryFiles(fileMask.str(), false);
+                sortDirectory(sortedFiles, *fileIterator, SD_bynameNC, sortInc, false); 
+                ForEachItemIn(sortedFilesIdx, sortedFiles)
+                {
+                    CDirectoryEntry &file = sortedFiles.item(sortedFilesIdx);
+                    StringBuffer output = file.name.get();
+                    if (!parseAndCheckWUID(mask, checkBeforeOrAfterWU, beforeWU, afterWU, output))
+                        continue;
+    
+                    const char *wuid = output.str();
+                    if (!conn)
+                        conn.setown(querySDS().connect(dfu ? "DFU/WorkUnits" : "WorkUnits", myProcessSession(), 0, 5*60*1000)); // connection to all
+                    if (conn->queryRoot()->hasProp(wuid)) // check not online
+                        continue;
+    
+                    Owned<IPropertyTree> wuTree;
+                    if (!isEmptyString(outputFormat) || !owner.isEmpty() || !state.isEmpty() || !cluster.isEmpty() ||
+                        !jobName.isEmpty() || !cmdName.isEmpty() || !priority.isEmpty() || !fileRead.isEmpty() ||
+                        !fileWritten.isEmpty() || !eclContains.isEmpty())
+                    {
+                        try
+                        {
+                            wuTree.setown(createPTree(*file.file));
+                            if (!wuTree || !checkFilters(wuTree))
+                                continue;
+                        }
+                        catch (IException *e)
+                        {
+                            StringBuffer msg;
+                            msg.appendf("WUiterate: Workunit %s failed to load", wuid);
+                            EXCLOG(e,msg.str());
+                            e->Release();
+                            continue;
+                        }
+                    }
+
+                    selectedWUIDs.append(wuid);
+                    if (!isEmptyString(outputFormat))
+                        selectedWUs.append(*wuTree.getClear());
+
+                    wuCount++;
+                    if (wuCount == numWUs)
+                        break;
+                }
+                if (wuCount == numWUs)
+                    break;
+            }
+            setOutput(cmd, outputFormat, selectedWUIDs, selectedWUs, sortInc);
```

Comment:
The loaded/deserialized wuids now in selectedWUs could be quite big.
Is there any reason you add all at end in setOutput?  i.e. why not get rid of the array an add them as you iterate ?